hbase-rsgroup/src/main/java/org/apache/hadoop/hbase/rsgroup/RSGroupAdminServer.java

```diff
@@ -541,6 +544,11 @@ public boolean balanceRSGroup(String groupName) throws IOException {
           getRSGroupAssignmentsByTable(master.getTableStateManager(), groupName);
       List<RegionPlan> plans = balancer.balanceCluster(assignmentsByTable);
       boolean balancerRan = !plans.isEmpty();
+
+      if (request.isDryRun()) {
+        return balancerRan;
```

Comment:
Is this correct? The balancer will never run when `dry_run=true`. Why does the response indicate otherwise? I guess this boolean response is not rich enough to communicate that a plan was calculated but not applied to the cluster.

Reply:
Ok so my initial thought here was to agree with you about the richness of the boolean, and to say that I didn't want to further mess with the interface to change the return value away from boolean. But I was just thinking, I'm already adding a new method that takes a BalanceRequest, so I could add a BalanceResponse. We can maintain backwards compatibility by making the default `balance()` and `balance(force)` methods call `response.isBalancerRan()` or something:

```java
default balance() throws IOException {
  return balance(BalanceRequest.defaultInstance()).isBalancerRan();
```

Something like that. 

I've addressed most of your feedback in one commit already. I'll try to take a stab at this now.

Reply:
Note: there's still some value in returning as I already have it implemented. The true value here does actually denote that the balancer ran -- it just didn't execute any moves. There are cases where this would be false, meaning the dry run literally didn't even try to execute. That would happen if there was already a balance in progress or if the cluster was in maintenance mode. 

I'm thinking a BalanceResponse could have three fields for now:

- boolean balancerRan (this would remain as today)
- int movesCalculated (this would be the same for dryrun vs execute)
- int movesExecuted  (this would be 0 for dryRun and probably equal to movesCalculated for execute)

I'll throw that up as a first pass and see what people think.

Reply:
I was anticipating a discussion, not code changes. I'm fine with leaving it as a boolean response, I just wanted to make sure we think about what the response values mean, and we document them.

There should already be a `BalancerResponse` in the protobuf, so there's a mechanism in place for adding a more rich response to the client experience. If we're adding to it, why not also include the amount of time spent on each calculation and execution of moves? I agree, it would be easy to add fields, and I'm not clear on the value they might provide.

Reply:
Ah ok. I'll leave this as-is for now. Thanks for the discussion!

I think we're in agreement on the value of "some non-bool return value" here. My only concern with leaving the boolean return value on my new `boolean balance(BalanceRequest request) throws IOException` (and async counterpart) method is if we'd have trouble modifying that interface in a follow-up PR. Any change to that method's return value will be a breaking change, so if we want to do that without further deprecations/gymnastics then we should at least put a placeholder now. I guess it would be fine as long as the follow-up comes before 2.5.0, right?

Reply:
Yes. I suggest we enrich this return type before 2.5.0. That can be this PR or a follow-on.